OmiseSDK/Client.swift

```diff
@@ -138,6 +138,113 @@ import os
         })
         dataTask.resume()
     }
+
+    public func retrieveChargeStatusWithCompletionHandler(from tokenID: String, completionHandler: (((ChargeStatus, Error?)) -> Void)?) {
+        let dataTask = session.dataTask(with: buildRetrieveTokenURLRquest(from: tokenID)) { (data, response, error) in
+            guard let completionHandler = completionHandler else { return } // nobody around to hear the leaf falls
+
+            var result: (ChargeStatus, Error?)
+            defer {
+                DispatchQueue.main.async {
+                    completionHandler(result)
+                }
+            }
+
+            if let error = error {
+                result = (.unknown, error)
+                return
+            }
+
+            guard let httpResponse = response as? HTTPURLResponse else {
+                let error = OmiseError.unexpected(error: .noErrorNorResponse, underlying: nil)
+                result = (.unknown, error)
+                return
+            }
+
+            let decoder = JSONDecoder()
+            decoder.dateDecodingStrategy = .formatted(Client.jsonDateFormatter)
+
+            switch httpResponse.statusCode {
+            case 400..<600:
+                guard let data = data else {
+                    let error = OmiseError.unexpected(error: .httpErrorWithNoData, underlying: nil)
+                    result = (.unknown, error)
+                    return
+                }
+
+                do {
+                    result = (.unknown, try decoder.decode(OmiseError.self, from: data))
+                } catch let err {
+                    let error = OmiseError.unexpected(error: .httpErrorResponseWithInvalidData, underlying: err)
+                    result = (.unknown, error)
+                }
+
+            case 200..<300:
+                guard let data = data else {
+                    let error = OmiseError.unexpected(error: .httpSucceessWithNoData, underlying: nil)
+                    result = (.unknown, error)
+                    return
+                }
+
+                do {
+                    let token = try decoder.decode(Token.self, from: data)
+                    result = (token.chargeStatus, nil)
+                } catch let err {
+                    let error = OmiseError.unexpected(error: .httpSucceessWithInvalidData, underlying: err)
+                    result = (.unknown, error)
+                }
+
+            default:
+                let error = OmiseError.unexpected(error: .unrecognizedHTTPStatusCode(code: httpResponse.statusCode), underlying: nil)
+                result = (.unknown, error)
+            }
+        }
+        dataTask.resume()
+    }
+    
+    public func pollingChargeStatusWithCompletionHandler(from tokenID: String, completionHandler: @escaping (ChargeStatus, Error?) -> Void) {
```

Comment:
What would you think about using `(Result<ChargeStatus, Error>) -> Void` instead of `(ChargeStatus, Error?) -> Void` as the completion?

Reply:
That good idea. Changed 🙇